src/stdlib/mod.rs

```diff
@@ -1,310 +1,80 @@
     lexer::Token,
     parser::{
-        types::{FnSig, GenericParameters, TyKind},
+        types::{FnSig, GenericParameters},
         ParserCtx,
     },
     type_checker::FnInfo,
-    var::{ConstOrCell, Value, Var},
+    var::Var,
 };
 
+pub mod bits;
+pub mod builtins;
 pub mod crypto;
 
-//
-// Builtins or utils (imported by default)
-// TODO: give a name that's useful for the user,
-//       not something descriptive internally like "builtins"
-
-pub const QUALIFIED_BUILTINS: &str = "std/builtins";
-
-const ASSERT_FN: &str = "assert(condition: Bool)";
-const ASSERT_EQ_FN: &str = "assert_eq(lhs: Field, rhs: Field)";
-const TO_BITS_FN: &str = "to_bits(const LEN: Field, val: Field) -> [Bool; LEN]";
-const FROM_BITS_FN: &str = "from_bits(bits: [Bool; LEN]) -> Field";
-
-// todo: each addition of builtins require changing this and `get_builtin_fn`,
-// - can we encapsulate them in a single place?
-/// List of builtin function signatures.
-pub const BUILTIN_SIGS: &[&str] = &[ASSERT_FN, ASSERT_EQ_FN, TO_BITS_FN, FROM_BITS_FN];
-
-// Unique set of builtin function names, derived from function signatures.
-pub static BUILTIN_FN_NAMES: Lazy<HashSet<String>> = Lazy::new(|| {
-    BUILTIN_SIGS
-        .iter()
-        .map(|s| {
-            let ctx = &mut ParserCtx::default();
-            let mut tokens = Token::parse(0, s).unwrap();
-            let sig = FnSig::parse(ctx, &mut tokens).unwrap();
-            sig.name.value
-        })
-        .collect()
-});
-
-pub fn get_builtin_fn<B: Backend>(name: &str) -> Option<FnInfo<B>> {
-    let ctx = &mut ParserCtx::default();
-    let mut tokens = Token::parse(0, name).unwrap();
-    let sig = FnSig::parse(ctx, &mut tokens).unwrap();
-
-    let fn_handle = match name {
-        ASSERT_FN => assert,
-        ASSERT_EQ_FN => assert_eq,
-        TO_BITS_FN => to_bits,
-        FROM_BITS_FN => from_bits,
-        _ => return None,
-    };
-
-    Some(FnInfo {
-        kind: FnKind::BuiltIn(sig, fn_handle),
-        span: Span::default(),
-    })
+pub enum AllStdModules {
+    Builtins,
+    Crypto,
+    Bits,
 }
 
-/// a function returns builtin functions
-pub fn builtin_fns<B: Backend>() -> Vec<FnInfo<B>> {
-    BUILTIN_SIGS
-        .iter()
-        .map(|sig| get_builtin_fn(sig).unwrap())
-        .collect()
-}
-
-/// Asserts that two vars are equal.
-fn assert_eq<B: Backend>(
-    compiler: &mut CircuitWriter<B>,
-    _generics: &GenericParameters,
-    vars: &[VarInfo<B::Field, B::Var>],
-    span: Span,
-) -> Result<Option<Var<B::Field, B::Var>>> {
-    // we get two vars
-    assert_eq!(vars.len(), 2);
-    let lhs_info = &vars[0];
-    let rhs_info = &vars[1];
-
-    // they are both of type field
-    if !matches!(lhs_info.typ, Some(TyKind::Field | TyKind::BigInt)) {
-        panic!(
-            "the lhs of assert_eq must be of type Field or BigInt. It was of type {:?}",
-            lhs_info.typ
-        );
+impl AllStdModules {
+    pub fn all_std_modules() -> Vec<AllStdModules> {
```

Comment:
is `new` better as a constructor name?

Reply:
Would using dynamic dispatch be better? like:

```rust
pub fn all_std_modules() -> Vec<Box<dyn Module>> {
    vec![
        Box::new(BuiltinsModule),
        Box::new(CryptoModule),
        Box::new(BitsModule),
    ]
}
```

But it can be tricky though. I remember it was difficult do something like this when it involves type parameters.

Reply:
I tried! but failed :D